pkg/services/authn/authnimpl/sync/user_sync.go

```diff
@@ -63,29 +80,88 @@ var (
 )
 
 func ProvideUserSync(userService user.Service, userProtectionService login.UserProtectionService, authInfoService login.AuthInfoService,
-	quotaService quota.Service, tracer tracing.Tracer, features featuremgmt.FeatureToggles,
+	quotaService quota.Service, tracer tracing.Tracer, features featuremgmt.FeatureToggles, cfg *setting.Cfg,
 ) *UserSync {
+	scimSection := cfg.Raw.Section("auth.scim")
 	return &UserSync{
-		userService:           userService,
-		authInfoService:       authInfoService,
-		userProtectionService: userProtectionService,
-		quotaService:          quotaService,
-		log:                   log.New("user.sync"),
-		tracer:                tracer,
-		features:              features,
-		lastSeenSF:            &singleflight.Group{},
+		allowedNonProvisionedUsers: scimSection.Key("allowed_non_provisioned_users").MustBool(false),
+		isUserProvisioningEnabled:  scimSection.Key("user_sync_enabled").MustBool(false),
+		userService:                userService,
+		authInfoService:            authInfoService,
+		userProtectionService:      userProtectionService,
+		quotaService:               quotaService,
+		log:                        log.New("user.sync"),
+		tracer:                     tracer,
+		features:                   features,
+		lastSeenSF:                 &singleflight.Group{},
 	}
 }
 
 type UserSync struct {
-	userService           user.Service
-	authInfoService       login.AuthInfoService
-	userProtectionService login.UserProtectionService
-	quotaService          quota.Service
-	log                   log.Logger
-	tracer                tracing.Tracer
-	features              featuremgmt.FeatureToggles
-	lastSeenSF            *singleflight.Group
+	allowedNonProvisionedUsers bool
+	isUserProvisioningEnabled  bool
+	userService                user.Service
+	authInfoService            login.AuthInfoService
+	userProtectionService      login.UserProtectionService
+	quotaService               quota.Service
+	log                        log.Logger
+	tracer                     tracing.Tracer
+	features                   featuremgmt.FeatureToggles
+	lastSeenSF                 *singleflight.Group
+}
+
+// ValidateUserProvisioningHook validates if a user should be allowed access based on provisioning status and configuration
+func (s *UserSync) ValidateUserProvisioningHook(ctx context.Context, id *authn.Identity, _ *authn.Request) error {
+	s.log.FromContext(ctx).Debug("Validating user provisioning", "auth_module", id.AuthenticatedBy, "auth_id", id.AuthID)
```

Comment:
You can create a local logger variable for `s.log.FromContext(ctx)`.

Reply:
+ you can add `auth_module` and `auth_id` fields to that logger and those two will be added to each log entry.

Reply:
I changed the code to reuse the same log object. I learned about reusing the same attributes, thank you for the suggestion.